tox.ini

```diff
@@ -1,2 +1,219 @@
+    pytest-cov
+    pytest-html
+    py{37,38,39,310,311}-test-{lin,win}: psycopg[binary]
+    test-mac: psycopg2-binary
+    py36-test: psycopg2-binary
+platform =
+    lin: linux
+    mac: darwin
+    win: win32
+allowlist_externals =
+    rm
+    {env:OPEN_CMD}
+
+[testenv:dep]
+description = Check package dependency problems
+commands = pipdeptree -w fail
+deps =
+    -r requirements.txt
+    pipdeptree
+
+[testenv:py{37,38,39,310,311}-type-{lin,mac,win}]
+description = Run static type checking with pyright
+labels =
+    type
+deps =
+    py{37,38,39,310,311}-type-{lin,win}: psycopg[binary]
+    type-mac: psycopg2-binary
+    py36-type: psycopg2-binary
+
+    -r requirements.txt
+    pyright
+commands = pyright --venv-path {toxworkdir}{/}{envname} --lib {posargs:patroni}
+platform =
+    lin: linux
+    mac: darwin
+    win: win32
+
+[testenv:black]
+description = Reformat code with black
+deps = black
+commands = black {posargs:patroni tests}
+
+[testenv:pg{12,13,14,15}-docker-build]
+description = Build docker containers needed for testing
+labels =
+    behave
+setenv =
+    pg11: PG_MAJOR = 11
+    pg12: PG_MAJOR = 12
+    pg13: PG_MAJOR = 13
+    pg14: PG_MAJOR = 14
+    pg15: PG_MAJOR = 15
+    DOCKER_BUILDKIT = 1
+commands =
+    docker build . \
+        --tag patroni-dev:{env:PG_MAJOR} \
+        --build-arg PG_MAJOR \
+        --file features/Dockerfile
+allowlist_externals = docker
+
+[testenv:pg{12,13,14,15}-docker-cmd]
+description = Run an arbitrary command in patroni-dev docker container
+setenv =
+    pg11: PG_MAJOR = 11
+    pg12: PG_MAJOR = 12
+    pg13: PG_MAJOR = 13
+    pg14: PG_MAJOR = 14
+    pg15: PG_MAJOR = 15
+commands =
+        docker run \
+        --env DCS=etcd \
+        --name tox-{env_name}-{env:PYTHONHASHSEED} \
+        --detach \
+        --rm \
+        patroni-dev:{env:PG_MAJOR} \
+        {posargs:echo "Hello world"}
+allowlist_externals = docker
+
+;[testenv:pg{11,12,13,14,15}-docker-behave-{etcd}-{lin,win,mac}]
+[testenv:pg{12,13,14,15}-docker-behave-{etcd}-{lin,mac}]
+description = Run behaviour tests in patroni-dev docker container
+setenv =
+    etcd: DCS=etcd
+    pg11: PG_MAJOR = 11
+    pg12: PG_MAJOR = 12
+    pg13: PG_MAJOR = 13
+    pg14: PG_MAJOR = 14
+    pg15: PG_MAJOR = 15
+    CONTAINER_NAME = tox-{env_name}-{env:PYTHONHASHSEED}
+labels =
+    behave
+depends =
+    pg{11,12,13,14,15}-docker-build
+
+# There's a bug which affects calling multiple envs on the command line
+# This should be a valid command: tox -e 'py{36,37,38,39,310,311}-behave-{env:DCS}-lin'
+# Replaced with workaround, see https://github.com/tox-dev/tox/issues/2850
+commands =
+    docker run \
+        --volume {tox_root}:/src \
+        --env DCS={env:DCS} \
+        --hostname {env:CONTAINER_NAME} \
+        --name {env:CONTAINER_NAME} \
+        --rm \
+        --tty \
+        {env:PATRONI_DEV_IMAGE:patroni-dev:{env:PG_MAJOR}} \
+        tox run -x 'tox.env_list=py\{36,37,38,39,310,311\}-behave-{env:DCS}-lin' \
```

Comment:
I have also spotted the following line in the tox docs

> tox works with the following Python interpreter implementations:
CPython versions 3.7, 3.8, 3.9, 3.10, 3.11

Reply:
> I don't think all these python interpreters are available in patroni-dev image

No, that's true. It's there in case a different image is provided or the upstream postgres image changes OS and `python-3` package becomes something else. It will skip any that don't exist. It also saves maintaining which is supported in the image if it needs to change during testing, e.g. I build a different image, or call the tox env directly instead of using the marker: `tox -e pg14-docker-behave-etcd-lin`. I also wanted to avoid making that env name even larger: `[testenv:py{36,37,38,39,310,311}-pg{12,13,14,15}-docker-behave-{etcd}-{lin,mac}]`!

Reply:
> I have also spotted the following line in the tox docs
> 
> > tox works with the following Python interpreter implementations:
> > CPython versions 3.7, 3.8, 3.9, 3.10, 3.11

Patroni still supports 3.6 doesn't it? I can remove it but it does appear to work though despite that.

Reply:
it is ok if you checked it works